back/app/services/projects_finder_service.rb

```diff
@@ -66,85 +136,55 @@ def followed_by_user
-  # in an active participatory phase (where user can probably do something),
-  # ordered by the end date of the current phase, soonest first (nulls last).
-  # Also returns action descriptors for each project, to avoid getting them again when serializing.
-  # => { projects: [Project], descriptor_pairs: { <project.id>: { <action_descriptors> }, ... } }
-  def participation_possible_uncached
-    subquery = @projects
+    base_scope = @projects
       .joins('INNER JOIN admin_publications AS admin_publications ON admin_publications.publication_id = projects.id')
-      .where(admin_publications: { publication_status: 'published' })
-
-    # Projects with active participatory (not information) phase & include the phases.end_at column
-    subquery = projects_with_active_phase(subquery)
-      .joins('INNER JOIN phases AS active_phases ON active_phases.project_id = projects.id')
-      .where.not(phases: { participation_method: 'information' })
-      .select('projects.*, projects.created_at AS projects_created_at, projects.id AS projects_id')
 
-    # Perform the SELECT DISTINCT on the outer query and order first by the end date of the active phase,
-    # second by project created_at, and third by project ID.
-    # Secondary & ternary orderings prevent duplicates when paginating, when prior ordering involves equivalent values
-    projects = Project
-      .from(subquery, :projects)
-      .distinct
-      .order('phase_end_at ASC NULLS LAST, projects_created_at ASC, projects_id ASC')
-      .preload(phases: { permissions: [:groups] })
-
-    # Projects user can participate in, or where such participation could (probably) be made possible by user
-    # (e.g. user not signed in).
-    # Unfortunately, this breaks the query chain, so we have to start a new one after this.
-    #
-    # Step 1: Create pairs of project ids and action descriptors.
-    # Since this is the last filtering step, we will keep going, until we reach the limit required for pagination.
-    pagination_limit = @page_size * @page_number
-    project_descriptor_pairs = {}
+    if @finished
+      finished_scope = base_scope.where(admin_publications: { publication_status: 'published' })
+      finished_scope = joins_last_phases_with_reports(finished_scope)
+        .where(
+          '(last_phases.last_phase_end_at < ? OR reports.id IS NOT NULL) AND admin_publications.publication_status = ?',
+          Time.zone.now, 'published'
+        )
+    end
 
-    projects.each do |project|
-      service = Permissions::ProjectPermissionsService.new(
-        project, @user, user_requirements_service: user_requirements_service
-      )
-      action_descriptors = service.action_descriptors
-      next unless service.participation_possible?(action_descriptors)
+    archived_scope = base_scope.where(admin_publications: { publication_status: 'archived' }) if @archived
+    archived_scope = joins_last_phases_with_reports(archived_scope) if @archived && @finished
```

Comment:
Do we need reports for archived projects? I guess we can still use it to add a 'Read the report' CTA... or did you have another purpose for it in mind?

Reply:
I think that I perform the same join with the archived projects as with the finished projects, in the case where we are filtering for both, so that the joins match. Without doing this, the SQL was invalid when I later order the collection, due to having different join structures - meaning it would fail with an error about incompatible joins.

Reply:
ChatGPT says:

> SQL incompatible joins occur when there are no matching records between tables based on the join condition, resulting in a join that produces either no rows or unexpected results. This often happens when using joins like `INNER JOIN` where only matching records are returned, so if no matches exist between the tables, the result set will be empty.

There's probably other ways I could have dealt with this issue, but I think that the resulting query will need to do something like my solution to make the tables compatible (archived vs finished), and it probably gets rationalised in the actual resulting query.

Reply:
But... this discussion has reminded me that I haven't preloaded the phase reports in the controller, making the join with reports less efficient.

I have just merged this PR, so I'll create another to add the preloading 🤦